### PR TITLE
add crud service and controller for user profile

### PR DESCRIPTION
**Pull request by SidletskyiOlexandr**

### CODE REVIEW

**Comment by Anatolii-Ivanov** [comment]
image

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Interfaces/IAzureBlobStorageService.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿using EasySpeak.Core.Common.DTO.UploadFile;
+
+namespace EasySpeak.Core.BLL.Interfaces
+{
+    public interface IAzureBlobStorageService
+    {
+        Task<UploadFileDto> AddFileAsync(NewFileDto newFileDto);
+        Task<UploadFileDto> GetFileAsync(long fileId);
+        Task<UploadFileDto> UpdateFileAsync(long fileId, NewFileDto newFileDto);
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Interfaces/IAzureBlobStorageService.cs:9]: UploadFile -> File/FileData/EasySpeakFile 
The same with UploadFile

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/AzureBlobStorageService.cs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            _blobServiceClient = blobServiceClient;
+        }
+
+        public async Task<UploadFileDto> AddFileAsync(NewFileDto newFileDto)
+        {
+            var path = await UploadFileAsync(newFileDto);
+
+            var uploadFile = new UploadFile()
+            {
+                CreatedAt = DateTime.UtcNow,
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/AzureBlobStorageService.cs:28]: We have CreatedAt date set automatically

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/AzureBlobStorageService.cs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                UpdatedAt = DateTime.UtcNow,
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/AzureBlobStorageService.cs:29]: We don't plan to update files, remove, create new

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/AzureBlobStorageService.cs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            if (file.Url != null)
+            {
+                await DeleteFromBlobAsync(file, newFileDto.FolderPath);
+            }
+
+            file.Url = pathToNewFile;
+            file.UpdatedAt = DateTime.UtcNow;
+            await _context.SaveChangesAsync();
+
+            return _mapper.Map<UploadFileDto>(file);
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/AzureBlobStorageService.cs:69]: Remove/create new

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/AzureBlobStorageService.cs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            if (blob.Exists())
+            {
+                return $"{Path.GetFileNameWithoutExtension(fileName)}_" +
+                    $"{DateTime.Now.ToString("yyyyMMddTHHmmssfff")}" +
+                    $"{Path.GetExtension(fileName)}";
+            }
+
+            return fileName;
+        }
+    }
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/AzureBlobStorageService.cs:142]: Move to separate service (file)

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/AzureBlobStorageService.cs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        {
+            if (newFileDto.Stream == null)
+            {
+                throw new Exception($"{newFileDto.FileName} is empty");
+            }
+
+            await CreateDirectoryAsync(newFileDto.FolderPath);
+            string uniqueFileName = CreateName(newFileDto.FileName, newFileDto.FolderPath);
+
+            var provider = new FileExtensionContentTypeProvider();
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/AzureBlobStorageService.cs:96]: Either var or types everywhere

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/AzureBlobStorageService.cs [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+            if (blob.Exists())
+            {
+                return $"{Path.GetFileNameWithoutExtension(fileName)}_" +
+                    $"{DateTime.Now.ToString("yyyyMMddTHHmmssfff")}" +
+                    $"{Path.GetExtension(fileName)}";
+            }
+
+            return fileName;
+        }
```
[the remark this answers is not in the document]
Comment (Anatolii-Ivanov) [reply backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/AzureBlobStorageService.cs:141]: +extension

File: backend/EasySpeak.Core/EasySpeak.Core.Common/DTO/UploadFile/NewFileDto.cs [hunk 8]
```diff
@@ -0,0 +1,9 @@
+﻿namespace EasySpeak.Core.Common.DTO.UploadFile
+{
+    public class NewFileDto
+    {
+        public Stream Stream { get; set; } = null!;
+        public string FolderPath { get; set; } = null!;
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.Common/DTO/UploadFile/NewFileDto.cs:6]: Won't be here, it's up to you set up folders

Comment (Anatolii-Ivanov) [reply backend/EasySpeak.Core/EasySpeak.Core.Common/DTO/UploadFile/NewFileDto.cs:6]: Generally we should have the same folder(blob container) for all profile pictures

File: backend/EasySpeak.Core/EasySpeak.Core.DAL/Context/EasySpeakCoreContext.cs [hunk 9]
```diff
@@ -19,7 +19,9 @@ public class EasySpeakCoreContext : DbContext
         public DbSet<Tag> Tags => Set<Tag>();
         public DbSet<Question> Questions => Set<Question>();
         public DbSet<Subquestion> Subquestions => Set<Subquestion>();
-        
+        public DbSet<UploadFile> UploadFiles => Set<UploadFile>();
+
+
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.DAL/Context/EasySpeakCoreContext.cs:24]: One new line

File: backend/EasySpeak.Core/EasySpeak.Core.DAL/Entities/UploadFile.cs [hunk 10]
```diff
@@ -0,0 +1,8 @@
+﻿namespace EasySpeak.Core.DAL.Entities
+{
+    public class UploadFile : Entity<long>
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.DAL/Entities/UploadFile.cs:3]: Not UploadFile, refer higher

File: backend/EasySpeak.Core/EasySpeak.Core.WebAPI/appsettings.json [hunk 11]
```diff
@@ -1,21 +1,24 @@
...
+  "FirebaseConfig": {
+    "project_id": "easy-meets-id-01",
+    "secure_uri": "https://securetoken.google.com"
+  },
+  "RabbitQueues": {
+    "NotificationQueue": "notifications"
+  },
+  "EmailSendURI": "https://localhost:5090/api/emailer/send",
+  "AzureBlobStorageSettings": {
+    "ConnectionString": "DefaultEndpointsProtocol=https;AccountName=easyspeak;AccountKey=7oWpdsuvuP952P0pIkSzqMD1xAq371tJJ9WHzhb+2vFC4sb/ckx9Fa1d5BcVGZ3WBlGMkAp61Kdu+AStPHATNg==;EndpointSuffix=core.windows.net"
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.WebAPI/appsettings.json:22]: Really bad idea

**Review by Anatolii-Ivanov — CHANGES_REQUESTED** [review]

File: backend/EasySpeak.Core/EasySpeak.Core.WebAPI/Controllers/UserProfileController.cs [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+        private readonly IFirebaseAuthService _firebaseAuthService;
+
+        public UserProfileController(IAzureBlobStorageService azureBlobStorageService, IFirebaseAuthService firebaseAuthService)
+        {
+            _azureBlobStorageService = azureBlobStorageService;
+            _firebaseAuthService = firebaseAuthService;
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> AddFile(IFormFile file)
```
[the remark this answers is not in the document]
Comment (SidletskyiOlexandr) [reply backend/EasySpeak.Core/EasySpeak.Core.WebAPI/Controllers/UserProfileController.cs]: In my task i should only add endpoint to update user profile photo without logic adding photo to user profile

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/AzureBlobStorageService.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Anatolii-Ivanov) [reply backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/AzureBlobStorageService.cs:77]: Not fixed

File: backend/EasySpeak.Core/EasySpeak.Core.WebAPI/appsettings.json [hunk 13]
```diff
@@ -1,21 +1,21 @@
...
+  },
+  "AllowedHosts": "*",
+  "FirebaseConfig": {
+    "project_id": "easy-meets-id-01",
+    "secure_uri": "https://securetoken.google.com"
+  },
+  "RabbitQueues": {
+    "NotificationQueue": "notifications"
+  },
+  "EmailSendURI": "https://localhost:5090/api/emailer/send"
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.WebAPI/appsettings.json:20]: Why you removed connection string name...

File: backend/EasySpeak.Core/EasySpeak.Core.WebAPI/Controllers/UserProfileController.cs [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+        [HttpPost]
+        public async Task<IActionResult> UploadAvatar(IFormFile file)
+        {
+            var fileDto = new NewEasySpeakFileDto()
+            {
+                Stream = file.OpenReadStream(),
+                FileName = file.FileName
+            };
+            var res = await _easySpeakFileService.AddFileAsync(fileDto);
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.WebAPI/Controllers/UserProfileController.cs:28]: Can you move this call to like UserService, where at the end this file will be assigned to user as a profile picture
Maybe you will need to create migration

Then call userService.UpdateProfilePhoto from this controller

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs [hunk 15]
```diff
@@ -53,5 +57,38 @@ public async Task<UserDto> GetUserAsync()
 
             return _mapper.Map<UserDto>(user);
         }
+
+        public async Task<UserProfilePhotoDto> UploadProfilePhoto(IFormFile file, long userId)
+        {
+            var user = await _context.Users.FirstOrDefaultAsync(u => u.Id == userId);
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs:63]: Please, use userId from firebaseAuthService

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs [hunk 16]
```diff
@@ -53,5 +57,38 @@ public async Task<UserDto> GetUserAsync()
...
+            }
+
+            user.ImageId = uploadFileDto.Id;
+            profilePhoto.UserId = userId;
+            await _context.SaveChangesAsync();
+
+            var photoDto = _mapper.Map<UserProfilePhotoDto>(user);
+            photoDto.PhotoUrl = uploadFileDto.Url;
+
+            return photoDto;
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs:91]: Returning just url will be fine

**Review by Anatolii-Ivanov — DISMISSED** [review]

**Review by potapchuk-anna — DISMISSED** [review]

**Review by Anatolii-Ivanov — APPROVED** [review]

**Review by potapchuk-anna — APPROVED** [review]

**Pull request merged**